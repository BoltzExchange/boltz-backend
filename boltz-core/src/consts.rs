pub const ECDSA_BYTES_TO_GRIND: usize = 2;

pub const PREIMAGE_DUMMY: [u8; 0] = [];

pub const STUB_SCHNORR_SIGNATURE_LENGTH: usize = 64;
