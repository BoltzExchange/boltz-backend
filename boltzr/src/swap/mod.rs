mod expiry_checker;
mod filters;
pub mod manager;
mod status;

pub use status::*;
