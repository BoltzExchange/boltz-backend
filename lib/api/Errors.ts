export default {
  COULD_NOT_PARSE_HEX: (argName: string): string =>
    `could not parse hex string: ${argName}`,
  INVALID_PARAMETER: (argName: string): string =>
    `invalid parameter: ${argName}`,
  UNDEFINED_PARAMETER: (argName: string): string =>
    `undefined parameter: ${argName}`,
  UNSUPPORTED_PARAMETER: (symbol: string, argName: string): string =>
    `${symbol} does not support ${argName}`,
  INVALID_SWAP_STATUS: (status: string): string =>
    `invalid swap status: ${status}`,
};
