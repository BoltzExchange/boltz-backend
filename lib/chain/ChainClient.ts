import BaseClient from '../BaseClient';
import Logger from '../Logger';
import RpcClient, { RpcConfig } from '../RpcClient';
import { ClientStatus } from '../consts/Enums';
import ChainClientInterface, { Info, Block, BestBlock } from './ChainClientInterface';

class ChainClient extends BaseClient implements ChainClientInterface {
  private rpcClient: RpcClient;
  private uri: string;

  constructor(private logger: Logger, config: RpcConfig, public readonly symbol: string) {
    super();

    this.rpcClient = new RpcClient(config);
    this.uri = `${config.host}:${config.port}`;

    this.bindWs();
  }

  private bindWs = () => {
    this.rpcClient.on('error', error => this.emit('error', error));
    this.rpcClient.on('message.orphan', (data) => {
      switch (data.method) {
        // Emits an event on mempool acceptance
        case 'relevanttxaccepted':
          data.params.forEach((transaction: string) => {
            this.emit('transaction.relevant.mempool', transaction);
          });
          break;

        // Emits an event when a blocks gets added
        case 'filteredblockconnected':
          const params: any[] = data.params;

          this.emit('block.connected', params[0]);

          if (params[2] !== null) {
            const transactions = params[2] as string[];

            transactions.forEach((transaction) => {
              this.emit('transaction.relevant.block', transaction, params[0]);
            });
          }
          break;
      }
    });
  }

  public connect = async () => {
    if (this.isDisconnected) {
      try {
        await this.rpcClient.connect();
        const info = await this.getInfo();

        if (info.version) {
          await this.notifyBlocks();

          this.clearReconnectTimer();
          this.setClientStatus(ClientStatus.Connected);
        } else {
          this.setClientStatus(ClientStatus.Disconnected);
          this.logger.error(`${this.symbol} at ${this.uri} is not able to connect, retrying in ${this.RECONNECT_INTERVAL} ms`);
          this.reconnectionTimer = setTimeout(this.connect, this.RECONNECT_INTERVAL);
        }
      } catch (error) {
        this.setClientStatus(ClientStatus.Disconnected);
        this.logger.error(`could not verify connection to chain ${this.symbol} chain at ${this.uri} because: ${JSON.stringify(error)}` +
        ` retrying in ${this.RECONNECT_INTERVAL} ms`);
        this.reconnectionTimer = setTimeout(this.connect, this.RECONNECT_INTERVAL);
      }
    }
  }

  public disconnect = async () => {
    this.clearReconnectTimer();
    this.setClientStatus(ClientStatus.Disconnected);

    await this.rpcClient.close();
  }

  public getInfo = (): Promise<Info> => {
    return this.rpcClient.call<Info>('getinfo');
  }

  public getBestBlock = (): Promise<BestBlock> => {
    return this.rpcClient.call<BestBlock>('getbestblock');
  }

  public getBlock = (blockHash: string): Promise<Block> => {
    return this.rpcClient.call<Block>('getblock', blockHash);
  }

  public loadTxFiler = (reload: boolean, addresses: string[], outpoints: string[]): Promise<null> => {
    // tslint:disable-next-line no-null-keyword
    return this.rpcClient.call<null>('loadtxfilter', reload, addresses, outpoints);
  }

  public sendRawTransaction = (rawTransaction: string, allowHighFees = true): Promise<string> => {
    return this.rpcClient.call<string>('sendrawtransaction', rawTransaction, allowHighFees);
  }

  public getRawTransaction = (transactionHash: string, verbose = 0) => {
    return this.rpcClient.call<any>('getrawtransaction', transactionHash, verbose);
  }

  public generate = (blocks: number): Promise<string[]> => {
    return this.rpcClient.call<string[]>('generate', blocks);
  }

  /**
   * Call this function to get notifications about the block acceptance of relevant transactions
   */
  private notifyBlocks = () => {
    return this.rpcClient.call<void>('notifyblocks');
  }

}

export default ChainClient;
