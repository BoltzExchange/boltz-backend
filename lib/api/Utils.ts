import { Request, Response } from 'express';
import Logger from '../Logger';
import { getHexBuffer } from '../Utils';
import MarkedSwapRepository from '../db/repositories/MarkedSwapRepository';
import CountryCodes from '../service/CountryCodes';
import ServiceErrors from '../service/Errors';
import Errors from './Errors';

type ApiArgument = {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'object';
  hex?: boolean;
  optional?: boolean;
};

// Some endpoints are getting spammed on mainnet, so we don't log the warnings for them
const errorsNotToLog: any[] = [
  ServiceErrors.SWAP_NO_LOCKUP().message,
  ServiceErrors.ETHEREUM_NOT_ENABLED().message,
];

/**
 * Validates that all required arguments were provided in the body correctly
 *
 * @returns the validated arguments
 */
export const validateRequest = (
  body: Record<string, any>,
  argsToCheck: ApiArgument[],
) => {
  const response: any = {};

  argsToCheck.forEach((arg) => {
    const value = body[arg.name];

    if (value !== undefined) {
      if (typeof value === arg.type) {
        if (arg.hex && value !== '') {
          const buffer = getHexBuffer(value);

          if (buffer.length === 0) {
            throw Errors.COULD_NOT_PARSE_HEX(arg.name);
          }

          response[arg.name] = buffer;
        } else {
          response[arg.name] = value;
        }
      } else {
        throw Errors.INVALID_PARAMETER(arg.name);
      }
    } else if (!arg.optional) {
      throw Errors.UNDEFINED_PARAMETER(arg.name);
    }
  });

  return response;
};

export const errorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  error: unknown,
  statusCode = 400,
): void => {
  if (typeof error === 'string') {
    writeErrorResponse(logger, req, res, statusCode, { error });
  } else {
    const errorObject = error as any;

    // Bitcoin Core related errors
    if (errorObject.details) {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.details,
      });
      // Custom error when broadcasting a refund transaction fails because
      // the locktime requirement has not been met yet
    } else if (errorObject.timeoutBlockHeight) {
      writeErrorResponse(logger, req, res, statusCode, error);
      // Everything else
    } else {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.message,
      });
    }
  }
};

export const successResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(200);

  if (typeof data === 'object') {
    res.json(data);
  } else {
    res.write(data);
    res.end();
  }
};

export const createdResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(201).json(data);
};

export const writeErrorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  statusCode: number,
  error: any,
) => {
  if (!errorsNotToLog.includes(error?.error || error)) {
    logger.warn(
      `Request ${req.method} ${req.originalUrl} ${
        req.body && Object.keys(req.body).length > 0
          ? `${JSON.stringify(req.body)} `
          : ''
      }failed: ${JSON.stringify(error)}`,
    );
  }

  setContentTypeJson(res);
  res.status(statusCode).json(error);
};

export const setContentTypeJson = (res: Response) => {
  res.set('Content-Type', 'application/json');
};

export const checkPreimageHashLength = (preimageHash: Buffer) => {
  if (preimageHash.length !== 32) {
    throw `invalid preimage hash length: ${preimageHash.length}`;
  }
};

export const markSwap = async (
  countryCodes: CountryCodes,
  ip: string,
  swapId: string,
) => {
  if (!countryCodes.isRelevantCountry(countryCodes.getCountryCode(ip))) {
    return;
  }

  await MarkedSwapRepository.addMarkedSwap(swapId);
};
