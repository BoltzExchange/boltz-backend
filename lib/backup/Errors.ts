import { concatErrorCode } from '../Utils';
import { ErrorCodePrefix } from '../consts/Enums';
import { Error } from '../consts/Types';

export default {
  BACKUP_DISABLED: (): Error => ({
    message: 'backups are disabled because of incomplete configuration',
    code: concatErrorCode(ErrorCodePrefix.Backup, 0),
  }),
};
