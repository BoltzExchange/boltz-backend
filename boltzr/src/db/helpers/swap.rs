use crate::db::helpers::{BoxedCondition, QueryResponse};
use crate::db::models::Swap;
use crate::db::schema::swaps;
use crate::db::Pool;
use crate::swap::SwapUpdate;
use diesel::prelude::*;
use diesel::{update, QueryDsl, RunQueryDsl, SelectableHelper};

pub type SwapCondition = BoxedCondition<swaps::table>;

pub trait SwapHelper {
    fn get_all(&self, condition: SwapCondition) -> QueryResponse<Vec<Swap>>;
    fn update_status(
        &self,
        id: &str,
        status: SwapUpdate,
        failure_reason: Option<String>,
    ) -> QueryResponse<usize>;
}

#[derive(Clone, Debug)]
pub struct SwapHelperDatabase {
    pool: Pool,
}

impl SwapHelperDatabase {
    pub fn new(pool: Pool) -> Self {
        Self { pool }
    }
}

impl SwapHelper for SwapHelperDatabase {
    fn get_all(&self, condition: BoxedCondition<swaps::dsl::swaps>) -> QueryResponse<Vec<Swap>> {
        Ok(swaps::dsl::swaps
            .select(Swap::as_select())
            .filter(condition)
            .load(&mut self.pool.get()?)?)
    }

    fn update_status(
        &self,
        id: &str,
        status: SwapUpdate,
        failure_reason: Option<String>,
    ) -> QueryResponse<usize> {
        if let Some(failure_reason) = failure_reason {
            Ok(update(swaps::dsl::swaps)
                .filter(swaps::dsl::id.eq(id))
                .set((
                    swaps::dsl::status.eq(status.to_string()),
                    swaps::dsl::failureReason.eq(failure_reason),
                ))
                .execute(&mut self.pool.get()?)?)
        } else {
            Ok(update(swaps::dsl::swaps)
                .filter(swaps::dsl::id.eq(id))
                .set(swaps::dsl::status.eq(status.to_string()))
                .execute(&mut self.pool.get()?)?)
        }
    }
}
