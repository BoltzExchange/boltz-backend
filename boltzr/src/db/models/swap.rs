use crate::db::models::{LightningSwap, SomeSwap, SwapType};
use crate::swap::SwapUpdate;
use crate::utils::pair::{split_pair, OrderSide};
use diesel::{AsChangeset, Insertable, Queryable, Selectable};

#[derive(Queryable, Selectable, Insertable, AsChangeset, PartialEq, Default, Clone, Debug)]
#[diesel(table_name = crate::db::schema::swaps)]
#[allow(non_snake_case)]
pub struct Swap {
    pub id: String,
    pub referral: Option<String>,
    pub pair: String,
    pub orderSide: i32,
    pub status: String,
    pub failureReason: Option<String>,
    pub invoice: Option<String>,
    pub lockupAddress: String,
    pub createdAt: chrono::NaiveDateTime,
}

impl SomeSwap for Swap {
    fn kind(&self) -> SwapType {
        SwapType::Submarine
    }

    fn id(&self) -> String {
        self.id.clone()
    }

    fn status(&self) -> SwapUpdate {
        SwapUpdate::parse(self.status.as_str())
    }
}

impl LightningSwap for Swap {
    fn chain_symbol(&self) -> anyhow::Result<String> {
        let pair = split_pair(&self.pair)?;
        Ok(if self.orderSide == OrderSide::Buy as i32 {
            pair.quote
        } else {
            pair.base
        })
    }

    fn lightning_symbol(&self) -> anyhow::Result<String> {
        let pair = split_pair(&self.pair)?;
        Ok(if self.orderSide == OrderSide::Buy as i32 {
            pair.base
        } else {
            pair.quote
        })
    }
}

#[cfg(test)]
mod test {
    use crate::db::models::{LightningSwap, SomeSwap, Swap, SwapType};
    use crate::swap::SwapUpdate;
    use crate::utils::pair::OrderSide;
    use rstest::*;

    #[test]
    fn test_kind() {
        assert_eq!(create_swap(None).kind(), SwapType::Submarine);
    }

    #[test]
    fn test_id() {
        let swap = create_swap(None);
        assert_eq!(swap.id(), swap.id);
    }

    #[test]
    fn test_status() {
        let swap = create_swap(None);
        assert_eq!(swap.status(), SwapUpdate::parse(swap.status.as_str()))
    }

    #[rstest]
    #[case(OrderSide::Buy, "BTC")]
    #[case(OrderSide::Sell, "L-BTC")]
    fn test_chain_symbol(#[case] side: OrderSide, #[case] expected: &str) {
        let swap = create_swap(Some(side));
        assert_eq!(swap.chain_symbol().unwrap(), expected);
    }

    #[rstest]
    #[case(OrderSide::Buy, "L-BTC")]
    #[case(OrderSide::Sell, "BTC")]
    fn test_lightning_symbol(#[case] side: OrderSide, #[case] expected: &str) {
        let swap = create_swap(Some(side));
        assert_eq!(swap.lightning_symbol().unwrap(), expected);
    }

    fn create_swap(order_side: Option<OrderSide>) -> Swap {
        Swap {
            id: "swap id".to_string(),
            pair: "L-BTC/BTC".to_string(),
            status: "transaction.mempool".to_string(),
            orderSide: order_side.unwrap_or(OrderSide::Buy) as i32,
            ..Default::default()
        }
    }
}
