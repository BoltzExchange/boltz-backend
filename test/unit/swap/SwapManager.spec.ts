import { ECPair, Network } from 'bitcoinjs-lib';
import { Networks, OutputType } from 'boltz-core';
import Logger from '../../../lib/Logger';
import Errors from '../../../lib/swap/Errors';
import Wallet from '../../../lib/wallet/Wallet';
import { getHexBuffer } from '../../../lib/Utils';
import { OrderSide } from '../../../lib/consts/Enums';
import LndClient from '../../../lib/lightning/LndClient';
import ChainClient from '../../../lib/chain/ChainClient';
import WalletManager from '../../../lib/wallet/WalletManager';
import { MinimalReverseSwapDetails } from '../../../lib/swap/SwapNursery';
import SwapManager, { SwapDetails, ReverseSwapDetails } from '../../../lib/swap/SwapManager';

const mockBindCurrency = jest.fn().mockImplementation();

const mockAddSwap = jest.fn().mockImplementation();
const mockAddReverseSwap = jest.fn().mockImplementation();

jest.mock('../../../lib/swap/SwapNursery', () => {
  return jest.fn().mockImplementation(() => ({
    addSwap: mockAddSwap,
    bindCurrency: mockBindCurrency,
    addReverseSwap: mockAddReverseSwap,
  }));
});

const blockchainInfo = {
  blocks: 123,
};
const mockGetBlockchainInfo = jest.fn().mockResolvedValue(blockchainInfo);

const mockUpdateOutputFilter = jest.fn().mockImplementation();

const mockSendRawTransaction = jest.fn().mockImplementation();

jest.mock('../../../lib/chain/ChainClient', () => {
  return jest.fn().mockImplementation(() => ({
    getBlockchainInfo: mockGetBlockchainInfo,
    updateOutputFilter: mockUpdateOutputFilter,
    sendRawTransaction: mockSendRawTransaction,
  }));
});

const mockedChainClient = <jest.Mock<ChainClient>><any>ChainClient;

const addedHoldInvoice = {
  paymentRequest: 'lnbcrt1hold',
};
const mockAddHoldInvoice = jest.fn().mockResolvedValue(addedHoldInvoice);

const decodedPayReq = {
  numSatoshis: 10000,
  routeHintsList: [],
  destination: '03605d88f1e91df6803a55f167d7e13f3dfa07254e9dd59840e3b818d91a9fe7ad',
  paymentHash: 'ddcdf979c7fde5c6fc775f38f1ff1b7dadf67deeddd9a7f4d77e5ef1af7af787b6ddc779d3cdf6e7bf1b77471a75c71b',
};
const mockDecodePayReq = jest.fn().mockImplementation(async (invoice: string) => {
  switch (invoice) {
    case 'noRoutes':
      return {
        ...decodedPayReq,
        destination: 'noRoutes',
      };

    case 'routeHints':
      return {
        ...decodedPayReq,
        destination: 'noRoutes',
        routeHintsList: [
          0,
        ],
      };

    default:
      return decodedPayReq;
  }
});

const mockQueryRoutes = jest.fn().mockImplementation(async (destination: string) => {
  switch (destination) {
    case 'noRoutes':
    case 'routeHints':
      return {
        routesList: [],
      };

    default:
      return {
        routesList: [
          0,
        ],
      };
  }
});

jest.mock('../../../lib/lightning/LndClient', () => {
  return jest.fn().mockImplementation(() => ({
    queryRoutes: mockQueryRoutes,
    decodePayReq: mockDecodePayReq,
    addHoldInvoice: mockAddHoldInvoice,
  }));
});

const mockedLndClient = <jest.Mock<LndClient>><any>LndClient;

const newKeys = {
  index: 42,
  keys: ECPair.fromWIF('cPWHfRKmx1JYxXxokET9rkQZxuMs3xTCqx9Vk9XDAgKZNVpFbTur', Networks.bitcoinRegtest),
};
const mockGetNewKeys = jest.fn().mockReturnValue(newKeys);

const encodedAddress = 'bcrt1';
const mockEncodeAddress = jest.fn().mockReturnValue(encodedAddress);

const transactionSent = {
  vout: 1,
  fee: 123,
  transaction: {
    getId: () => 'id',
    toHex: () => 'hex',
    getHash: () => getHexBuffer('0xfff'),
  },
};
const mockSendToAddress = jest.fn().mockResolvedValue(transactionSent);

jest.mock('../../../lib/wallet/Wallet', () => {
  return jest.fn().mockImplementation(() => ({
    getNewKeys: mockGetNewKeys,
    encodeAddress: mockEncodeAddress,
    sendToAddress: mockSendToAddress,
  }));
});

const mockedWallet = <jest.Mock<Wallet>><any>Wallet;

jest.mock('../../../lib/wallet/WalletManager', () => {
  return jest.fn().mockImplementation(() => ({
    wallets: new Map<string, Wallet>([
      ['BTC', mockedWallet()],
      ['LTC', mockedWallet()],
    ]),
  }));
});

const mockedWalletManager = <jest.Mock<WalletManager>><any>WalletManager;

describe('SwapManager', () => {
  const currencies = [
    {
      symbol: 'BTC',
      network: Networks.bitcoinRegtest,
      config: {} as any,
      lndClient: mockedLndClient(),
      chainClient: mockedChainClient(),
    },
    {
      symbol: 'LTC',
      network: Networks.litecoinRegtest,
      config: {} as any,
      lndClient: mockedLndClient(),
      chainClient: mockedChainClient(),
    },
    {
      symbol: 'DOGE',
      network: Networks.dogecoinRegtest as Network,
      config: {} as any,
      chainClient: mockedChainClient(),
    },
  ];

  const manager = new SwapManager(
    Logger.disabledLogger,
    mockedWalletManager(),
    currencies,
  );

  beforeEach(() => {
    mockGetNewKeys.mockClear();
    mockEncodeAddress.mockClear();
    mockGetBlockchainInfo.mockClear();
    mockUpdateOutputFilter.mockClear();
  });

  test('should create swaps', async () => {
    const invoice = 'lnbcrt1';
    const expectedAmount = 123;
    const baseCurrency = 'LTC';
    const quoteCurrency = 'BTC';
    const acceptZeroConf = true;
    const timeoutBlockDelta = 2;
    const orderSide = OrderSide.BUY;
    const outputType = OutputType.Compatibility;
    const refundPublicKey = getHexBuffer('0xfff');

    const expectedOutputScript = getHexBuffer('a914205d5fb9e785f0df21c9fac5594496e90135058387');
    const expectedTimeoutBlockHeight = blockchainInfo.blocks + timeoutBlockDelta;

    const response = await manager.createSwap(
      baseCurrency,
      quoteCurrency,
      orderSide,
      invoice,
      expectedAmount,
      refundPublicKey,
      outputType,
      timeoutBlockDelta,
      acceptZeroConf,
    );

    expect(response).toEqual({
      address: encodedAddress,
      keyIndex: newKeys.index,
      timeoutBlockHeight: expectedTimeoutBlockHeight,

      // tslint:disable-next-line: max-line-length
      redeemScript: 'a914c7a349a62d13115097a38475d4002b67084e0b0f876321023d9e44575cd6f03dbc851cc9a6e037339302d0faad6a8d86284745146850633e67017db1750068ac',
    });

    expect(mockGetBlockchainInfo).toHaveBeenCalledTimes(1);

    expect(mockDecodePayReq).toHaveBeenCalledTimes(1);
    expect(mockDecodePayReq).toHaveBeenCalledWith(invoice);

    expect(mockQueryRoutes).toHaveBeenCalledTimes(1);
    expect(mockQueryRoutes).toHaveBeenCalledWith(
      decodedPayReq.destination,
      decodedPayReq.numSatoshis,
    );

    expect(mockGetNewKeys).toHaveBeenCalledTimes(1);

    expect(mockAddSwap).toHaveBeenCalledTimes(1);
    expect(mockAddSwap).toHaveBeenCalledWith(
      {
        ...currencies[0],
        wallet: expect.anything(),

        swaps: expect.anything(),
        swapTimeouts: expect.anything(),

        reverseSwaps: expect.anything(),
        reverseSwapTimeouts: expect.anything(),
        reverseSwapTransactions: expect.anything(),
      },
      {
        invoice,
        outputType,
        expectedAmount,
        acceptZeroConf,
        claimKeys: newKeys.keys,
        sendingSymbol: baseCurrency,
        redeemScript: getHexBuffer(response.redeemScript),
      },
      getHexBuffer('a914205d5fb9e785f0df21c9fac5594496e90135058387'),
      expectedTimeoutBlockHeight,
    );

    expect(mockUpdateOutputFilter).toHaveBeenCalledTimes(1);
    expect(mockUpdateOutputFilter).toHaveBeenCalledWith([
      expectedOutputScript,
    ]);

    expect(mockEncodeAddress).toHaveBeenCalledTimes(1);
    expect(mockEncodeAddress).toHaveBeenCalledWith(expectedOutputScript);

    // Should not perform a route check if the invoice has route hints
    await expect(manager.createSwap(
      baseCurrency,
      quoteCurrency,
      orderSide,
      'routeHints',
      expectedAmount,
      refundPublicKey,
      outputType,
      timeoutBlockDelta,
      acceptZeroConf,
    )).resolves.toEqual(response);

    // Check that no routes were queried
    expect(mockQueryRoutes).toHaveBeenCalledTimes(1);

    // Throw if no route can be found
    await expect(manager.createSwap(
      baseCurrency,
      quoteCurrency,
      orderSide,
      'noRoutes',
      expectedAmount,
      refundPublicKey,
      outputType,
      timeoutBlockDelta,
      acceptZeroConf,
    )).rejects.toEqual(Errors.NO_ROUTE_FOUND());

    // Throw if the lightning currency doesn't have a LND client
    await expect(manager.createSwap(
      'DOGE',
      quoteCurrency,
      orderSide,
      '',
      expectedAmount,
      refundPublicKey,
      outputType,
      timeoutBlockDelta,
      acceptZeroConf,
    )).rejects.toEqual(Errors.NO_LND_CLIENT('DOGE'));
  });

  test('should create reverse swaps', async () => {
    const invoiceAmont = 246;
    const onchainAmount = 123;
    const baseCurrency = 'LTC';
    const quoteCurrency = 'BTC';
    const timeoutBlockDelta = 2;
    const orderSide = OrderSide.BUY;
    const outputType = OutputType.Bech32;
    const claimPublicKey = getHexBuffer('0xfff');
    const preimageHash = getHexBuffer('083236df9ca7cffc49a5164167595ae3f77f17b3fc81ed6530134816a29f9a14');

    // tslint:disable-next-line: max-line-length
    const expectedRedeemScript = '8201208763a9145c44e40d43303607cabd1dd1fd00384ac41c6d5d88006775017db17521023d9e44575cd6f03dbc851cc9a6e037339302d0faad6a8d86284745146850633e68ac';
    const expectedOutputScript = getHexBuffer('00204ad5075f8a83a2736ebca13fc319702e5bc3e274a9cd579da2da0dba183175b5');

    const response = await manager.createReverseSwap(
      baseCurrency,
      quoteCurrency,
      orderSide,
      preimageHash,
      invoiceAmont,
      onchainAmount,
      claimPublicKey,
      outputType,
      timeoutBlockDelta,
    );

    expect(response).toEqual({
      keyIndex: newKeys.index,
      lockupAddress: encodedAddress,
      redeemScript: expectedRedeemScript,
      invoice: addedHoldInvoice.paymentRequest,
      timeoutBlockHeight: timeoutBlockDelta + blockchainInfo.blocks,
    });

    expect(mockAddHoldInvoice).toHaveBeenCalledTimes(1);
    expect(mockAddHoldInvoice).toHaveBeenCalledWith(invoiceAmont, preimageHash, 'Send to LTC address');

    expect(mockGetNewKeys).toHaveBeenCalledTimes(1);

    expect(mockGetBlockchainInfo).toHaveBeenCalledTimes(1);

    expect(mockEncodeAddress).toHaveBeenCalledTimes(1);
    expect(mockEncodeAddress).toHaveBeenCalledWith(expectedOutputScript);

    expect(mockUpdateOutputFilter).toHaveBeenCalledTimes(1);
    expect(mockUpdateOutputFilter).toHaveBeenCalledWith([expectedOutputScript]);

    expect(mockAddReverseSwap).toHaveBeenCalledTimes(1);
    expect(mockAddReverseSwap).toHaveBeenCalledWith(
      {
        outputType,
        preimageHash,
        refundKeys: newKeys.keys,
        sendingSymbol: baseCurrency,
        receivingSymbol: quoteCurrency,
        redeemScript: getHexBuffer(expectedRedeemScript),

        sendingDetails: {
          amount: onchainAmount,
          address: encodedAddress,
        },
      },
      addedHoldInvoice.paymentRequest,
      response.timeoutBlockHeight,
    );
  });

  test('should check routability', async () => {
    const checkRoutability = manager['checkRoutability'];

    const lndClient = mockedLndClient();

    await expect(checkRoutability(lndClient, '', 1)).resolves.toBeTruthy();
    await expect(checkRoutability(lndClient, 'noRoutes', 1)).resolves.toBeFalsy();

    // Return false if "queryRoutes" throws
    lndClient['queryRoutes'] = async () => {
      throw 'whatever';
    };

    await expect(checkRoutability(lndClient, '', 1)).resolves.toBeFalsy();
  });

  test('should get currencies', () => {
    const getCurrencies = manager['getCurrencies'];

    expect(getCurrencies('LTC', 'BTC', OrderSide.BUY)).toEqual({
      sendingCurrency: {
        ...manager['currencies'].get('LTC'),
        wallet: manager['walletManager'].wallets.get('LTC'),
      },
      receivingCurrency: {
        ...manager['currencies'].get('BTC'),
        wallet: manager['walletManager'].wallets.get('BTC'),
      },
    });

    expect(getCurrencies('LTC', 'BTC', OrderSide.SELL)).toEqual({
      sendingCurrency: {
        ...manager['currencies'].get('BTC'),
        wallet: manager['walletManager'].wallets.get('BTC'),
      },
      receivingCurrency: {
        ...manager['currencies'].get('LTC'),
        wallet: manager['walletManager'].wallets.get('LTC'),
      },
    });

  });

  test('should get currency', () => {
    const getCurrency = manager['getCurrency'];

    expect(getCurrency('BTC')).toEqual({
      ...currencies[0],

      swaps: new Map<string, SwapDetails>(),
      swapTimeouts: new Map<number, string[]>(),

      reverseSwaps: new Map<string, ReverseSwapDetails>(),
      reverseSwapTransactions: new Map<string, MinimalReverseSwapDetails>(),
      reverseSwapTimeouts: new Map<number, MinimalReverseSwapDetails[]>(),
    });

    // Throw if currency cannot be found
    const notFound = 'notFound';

    expect(() => getCurrency(notFound)).toThrow(
      Errors.CURRENCY_NOT_FOUND(notFound).message ,
    );
  });
});
