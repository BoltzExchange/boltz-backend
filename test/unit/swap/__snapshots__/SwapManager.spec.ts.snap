// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapManager should create reverse swap with covenant 1`] = `
{
  "claimLeaf": {
    "output": "82012088a9148265576ce381e50302d7b2afb7e571fda90e6dc4882002804e7f86e9ca29f582f1fd2b91e6eee6be10b5e2b086dfa52a14aa8ca63fcbac",
    "version": 196,
  },
  "covenantClaimLeaf": {
    "output": "82012088a9148265576ce381e50302d7b2afb7e571fda90e6dc48800d10088142606412ac0eebf1b779cf5b0e52578315b6a19d28800ce51882025b251070e29ca19043cf33ccd7324e2ddab03ecc4ae0b5e77c4fc0e5cf6c95a8800cf7508a40100000000000087",
    "version": 196,
  },
  "refundLeaf": {
    "output": "20c9c71ee3fee0c400ff64e51e955313e77ea499fc609973c71c5a4104a8d903bbad02f600b1",
    "version": 196,
  },
}
`;
